misc-scripts/xref_mapping/XrefParser/UniProtParser/Loader.pm

```diff
@@ -0,0 +1,122 @@
+=head1 LICENSE
+
+Copyright [1999-2015] Wellcome Trust Sanger Institute and the EMBL-European Bioinformatics Institute
+Copyright [2016-2018] EMBL-European Bioinformatics Institute
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+
+
+package XrefParser::UniProtParser::Loader;
+
+use strict;
+use warnings;
+
+use Carp;
+
+
+sub new {
+  my ( $proto, $arg_ref ) = @_;
+
+  my $self = {
+              'batch_size'  => $arg_ref->{'batch_size'} // 1,
+              'dbh'         => $arg_ref->{'dbh'},
+              '_baseParser' => $arg_ref->{'baseParser'},
+            };
+  my $class = ref $proto || $proto;
+  bless $self, $class;
+  $self->_clear_send_buffer();
+
+  return $self;
+}
+
+
+sub DESTROY {
+  my ( $self ) = @_;
+
+  $self->flush();
+  return;
+}
+
+
+sub baseParserInstance {
+  my ( $self ) = @_;
+  return $self->{'_baseParser'};
+}
+
+
+sub load {
+  my ( $self, $transformed_data ) = @_;
+
+  if ( ! defined $transformed_data ) {
+    return;
+  }
+
+  $self->_add_to_send_buffer( $transformed_data );
+
+  if ( $self->{'send_backlog'} >= $self->{'batch_size'} ) {
+    $self->flush();
+  }
+
+  return;
+}
+
+
+sub flush {
+  my ( $self ) = @_;
+
+  my $bp = $self->baseParserInstance();
+
+  if ( ! $bp->upload_xref_object_graphs($self->{'send_buffer'},
+                                        $self->{'dbh'}) ) {
+    croak 'Failed to upload xref object graphs. Check for errors on STDOUT)';
+  }
+
+  $self->_clear_send_buffer();
+
+  return;
+}
+
+
+sub _add_to_send_buffer {
+  my ( $self, $entry ) = @_;
+
+  # Play it safe and assume that parser-private fields *could* pollute
+  # the output unless removed from it
+  my ( $multiplicity ) = delete $entry->{'_multiplicity'};
```

Comment:
I know a single UniProt record can generate multiple xrefs, because there can be a few or no dependent xrefs, and that this varies per record, but I'm not sure why you care about smoothing it out enough to add this construct. Batching at a level higher, pushing e.g. a 1000 xrefs and all their dependent xrefs, will probably work just as well.

Reply:
Looking at the species_id-taxonomy_id map in the Ensembl database, for certain species the number of generated xrefs can be several times that of the number of parsed UniProt records. I could live with the loader inserting e.g. 1200 xrefs instead of 1000 (the batch size is still a thousand) but expecting 1000 and inserting _e.g._ 3000 would be a fairly sizeable discrepancy. Therefore, with multiplicity counting having been implemented in the old UniProtParser I saw no reason not to implement it in the new version.

Reply:
You're now getting it from `$self->_recognised_taxon_ids();` in the Transformer. It's some per-species value that I think doesn't achieve what you want.

If this field is meant to be about evening out the payload, you could count the number of dependent xrefs in the loader - 1 plus however many dependent xrefs it has.